### ISSUE TITLE
[testing branch] Startup error on bots with existing configurations, related to channel ban databases

### ISSUE THREAD

**Issue by GLolol** [ProgVal/Limnoria#1106]
Changes made in 6efea561a57a64bf91c61a5d1f3243db7e26f4f3 appears to have broken the ban database for existing bots.

Error on start:

```python
gl@midnight:~/limnoria$ supybot testbot.conf
ERROR 2015-05-15T17:51:58 Invalid channel database, resetting to empty.
ERROR 2015-05-15T17:51:58 Exact error:
Traceback (most recent call last):
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 835, in open
    self.flush()
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 853, in flush
    c.preserve(fd, indent='  ')
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 481, in preserve
    utils.sortBy(lambda x:x[1][0], bans)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/utils/gen.py", line 143, in sortBy
    L[i] = (f(elt), i, elt)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 481, in <lambda>
    utils.sortBy(lambda x:x[1][0], bans)
TypeError: 'int' object is not subscriptable
```

Similar error on shutdown:

```python
q^CINFO 2015-05-15T17:52:12 Exiting due to Ctrl-C.  If the bot doesn't exit within a few seconds, feel free to press Ctrl-C again to make it exit without flushing its message queues.
ERROR 2015-05-15T17:52:12 Uncaught exception in flusher #4 (<bound method ChannelsDictionary.flush of <supybot.ircdb.ChannelsDictionary object at 0x7f23818cd748>>):
Traceback (most recent call last):
  File "/home/gl/.local/bin/supybot", line 90, in main
    drivers.run()
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/drivers/__init__.py", line 107, in run
    log.exception('Uncaught exception in in drivers.run:')
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/log.py", line 80, in exception
    logging.Logger.exception(self, *args)
  File "/usr/lib/python3.4/logging/__init__.py", line 1315, in exception
    self.error(msg, *args, **kwargs)
  File "/usr/lib/python3.4/logging/__init__.py", line 1308, in error
    self._log(ERROR, msg, args, **kwargs)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/log.py", line 89, in _log
    extra=extra)
  File "/usr/lib/python3.4/logging/__init__.py", line 1414, in _log
    self.handle(record)
  File "/usr/lib/python3.4/logging/__init__.py", line 1424, in handle
    self.callHandlers(record)
  File "/usr/lib/python3.4/logging/__init__.py", line 1486, in callHandlers
    hdlr.handle(record)
  File "/usr/lib/python3.4/logging/__init__.py", line 853, in handle
    self.emit(record)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/log.py", line 128, in emit
    msg = self.format(record)
  File "/usr/lib/python3.4/logging/__init__.py", line 828, in format
    return fmt.format(record)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/log.py", line 66, in format
    return logging.Formatter.format(self, record)
  File "/usr/lib/python3.4/logging/__init__.py", line 573, in format
    record.exc_text = self.formatException(record.exc_info)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/drivers/__init__.py", line 105, in run
    driver.run()
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/drivers/Socket.py", line 187, in run
    self._select()
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/drivers/Socket.py", line 160, in _select
    [], [], conf.supybot.drivers.poll())
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/world.py", line 120, in flush
    f()
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 853, in flush
    c.preserve(fd, indent='  ')
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 481, in preserve
    utils.sortBy(lambda x:x[1][0], bans)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/utils/gen.py", line 143, in sortBy
    L[i] = (f(elt), i, elt)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 481, in <lambda>
    utils.sortBy(lambda x:x[1][0], bans)
TypeError: 'int' object is not subscriptable
ERROR 2015-05-15T17:52:12 Exception id: 0x8cf5d
INFO 2015-05-15T17:52:12 Flushers flushed and garbage collected.
INFO 2015-05-15T17:52:12 Driver for Irc object for overdrive-irc dying.
INFO 2015-05-15T17:52:12 Irc object for overdrive-irc dying.
INFO 2015-05-15T17:52:12 Driver for Irc object for overdrive-irc dying.
INFO 2015-05-15T17:52:12 Removing driver SocketDriver(Irc object for overdrive-irc).
INFO 2015-05-15T17:52:12 Total uptime: 14 seconds.
INFO 2015-05-15T17:52:12 Total CPU time taken: 1.17 seconds.
INFO 2015-05-15T17:52:12 No more Irc objects, exiting.
INFO 2015-05-15T17:52:12 Shutdown initiated.
INFO 2015-05-15T17:52:12 Killing Driver objects.
INFO 2015-05-15T17:52:12 Killing Irc objects.
INFO 2015-05-15T17:52:12 Writing registry file to testbot.conf
INFO 2015-05-15T17:52:12 Finished writing registry file.
ERROR 2015-05-15T17:52:12 Uncaught exception in flusher #4 (<bound method ChannelsDictionary.flush of <supybot.ircdb.ChannelsDictionary object at 0x7f23818cd748>>):
Traceback (most recent call last):
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/world.py", line 120, in flush
    f()
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 853, in flush
    c.preserve(fd, indent='  ')
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 481, in preserve
    utils.sortBy(lambda x:x[1][0], bans)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/utils/gen.py", line 143, in sortBy
    L[i] = (f(elt), i, elt)
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 481, in <lambda>
    utils.sortBy(lambda x:x[1][0], bans)
TypeError: 'int' object is not subscriptable
ERROR 2015-05-15T17:52:12 Exception id: 0x8cf5d
INFO 2015-05-15T17:52:12 Shutdown complete.
```

**Comment by ProgVal** [comment]
Could you try this fix?

```
diff --git a/src/ircdb.py b/src/ircdb.py
index 03dcd4b..00c58c0 100644
--- a/src/ircdb.py
+++ b/src/ircdb.py
@@ -478,7 +478,7 @@ class IrcChannel(object):
         for capability in self.capabilities:
             write('capability ' + capability)
         bans = self.bans.items()
-        utils.sortBy(lambda x:x[1][0], bans)
+        utils.sortBy(lambda x:x[1], bans)
         for (ban, (expiration, description)) in bans:
             write('ban %s %d %s' % (ban, expiration, description))
         ignores = self.ignores.items()
```

**Comment by GLolol** [comment]
```
gl@midnight:~/limnoria$ supybot testbot.conf
ERROR 2015-05-15T18:48:14 Invalid channel database, resetting to empty.
ERROR 2015-05-15T18:48:14 Exact error:
Traceback (most recent call last):
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 835, in open
    self.flush()
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 853, in flush
    c.preserve(fd, indent='  ')
  File "/home/gl/.local/lib/python3.4/site-packages/supybot/ircdb.py", line 482, in preserve
    for (ban, (expiration, description)) in bans:
TypeError: 'int' object is not iterable
ERROR 2015-05-15T18:48:14 Exception id: 0xce53b
```

**Comment by ProgVal** [comment]
grmbl

**Comment by ProgVal** [comment]
What about this instead?

```diff
diff --git a/src/ircdb.py b/src/ircdb.py
index 03dcd4b..d785f74 100644
--- a/src/ircdb.py
+++ b/src/ircdb.py
@@ -478,6 +478,8 @@ class IrcChannel(object):
         for capability in self.capabilities:
             write('capability ' + capability)
         bans = self.bans.items()
+        bans = [(x, (y, None) if isinstance(y, int) else y)
+                for (x, y) in bans]
         utils.sortBy(lambda x:x[1][0], bans)
         for (ban, (expiration, description)) in bans:
             write('ban %s %d %s' % (ban, expiration, description))
```

**Comment by GLolol** [comment]
Yup, the last one works. :smiley:

**Comment by GLolol** [comment]
Fixed by f1fae13b4a292885a59988da66cf34c32ce8e351 - closing as the bug only affects testing, not master.
